make `GILOnceCell::get_or_try_init_type_ref` public

glevco opened the pull request:
Resolves https://github.com/PyO3/pyo3/issues/4516.

- Change `GILOnceCell::get_or_try_init_type_ref` to `GILOnceCell::import` and make it public API, adding an example for it.
- Update "Executing existing Python code" section of the guide:
  - Add mention to `GILOnceCell::import`.
  - Change `Python::eval_bound` (which is deprecated) to `Python::eval`.
  - Change `Python::run_bound` (which is deprecated) to `Python::run`, also adding missing link.

mejrs commented on src/sync.rs:
````diff
@@ -218,7 +218,35 @@ impl GILOnceCell<Py<PyType>> {
+    /// # });
+    /// ```
+    ///
+    pub fn get_or_try_init_type_ref<'py>(
````
This can use a better name. Maybe just "import"? 🤔

glevco replied:
I agree, but then maybe it should be implemented for `GILOnceCell<Py<T>> where T: PyTypeInfo` instead of `GILOnceCell<Py<PyType>>`? Otherwise, I would call it `import_type`.

davidhewitt replied:
I'd be fine with that. I think you want `T: PyTypeCheck` rather than `T: PyTypeInfo`, though.

glevco replied:
Right, done in https://github.com/PyO3/pyo3/pull/4542/commits/79568f6c267d8973504eb94b19ae0f32a92d066c!

mejrs reviewed (commented):
> Change `PyModule::import` suggestion to `Python::import` instead

Why? I don't have a strong opinion either way, but slightly prefer `PyModule::import`.

glevco commented:
@mejrs 

> Why? I don't have a strong opinion either way, but slightly prefer `PyModule::import`.

Since `Python::import` was used in the implementation of `get_or_try_init_type_ref`, I assumed it was preferable. Also, it pairs nicely with the other entries in this section of the guide, as they recommend other "global" functions such as `Python::eval` and `Python::run`. What do you think?

davidhewitt commented:
I prefer `PyModule::import` too, I think it's nicer to think of that as a "constructor" for the `PyModule` type.

mejrs commented:
> What do you think?

From a documentation perspective I would prefer that people visit `PyModule::import` over `Python::import`. `Python` has a lot of not module related api, and its page is a lot bigger.

Also, using "constructor" syntax also makes it clear that when they use it, they're getting a module (rather than something arbitrary like `from foo import bar` might do).

glevco commented:
@davidhewitt @mejrs Fair enough, I reverted the guide back to `PyModule::import` in https://github.com/PyO3/pyo3/pull/4542/commits/79568f6c267d8973504eb94b19ae0f32a92d066c. I also changed the method name to `import` and made its impl generic for `PyTypeCheck`, as discussed above. I updated the PR description accordingly.

mejrs commented on src/sync.rs:
```diff
@@ -214,16 +214,46 @@ impl<T> GILOnceCell<Py<T>> {
     /// This is a shorthand method for `get_or_init` which imports the type from Python on init.
-    pub(crate) fn get_or_try_init_type_ref<'py>(
+    ///
+    /// # Example: Using `GILOnceCell` to avoid the overhead of importing a class multiple times
+    ///
```
It is better to just say what the example does, this makes for nicer URLs as well.

```suggestion
    /// # Example: Using `GILOnceCell` to store a class in a static variable.
    ///
    /// `GILOnceCell` can be used to avoid importing a class multiple times
```

(this also explicitly mentions "static", which is important. We don't want people to have local (or worse, `const`) GILOnceCells).

glevco replied:
Done: https://github.com/PyO3/pyo3/pull/4542/commits/5d95fbad3bac21a900936eb65cc20fb076751cf8

mejrs reviewed (commented):
Sorry for the late review. I have a couple of nits about the documentation.

glevco commented on src/sync.rs:
```diff
@@ -214,16 +214,46 @@ impl<T> GILOnceCell<Py<T>> {
+where
+    T: PyTypeCheck,
+{
     /// Get a reference to the contained Python type, initializing it if needed.
```
Done! https://github.com/PyO3/pyo3/pull/4542/commits/bcb883c106b4f8067ba281aa7aa2ee658cd4ba00

glevco commented on guide/src/python-from-rust/calling-existing-code.md:
```diff
@@ -24,9 +23,16 @@ fn main() -> PyResult<()> {
+[`PyModule::import`] introduces an overhead each time it's called. To avoid this in functions that are called multiple times,
+using a [`GILOnceCell`]({{#PYO3_DOCS_URL}}/pyo3/sync/struct.GILOnceCell.html) is recommended. Specifically, for importing types,
+[`GILOnceCell::import`]({{#PYO3_DOCS_URL}}/pyo3/sync/struct.GILOnceCell.html#method.import) is provided
+(check out the [example]({{#PYO3_DOCS_URL}}/pyo3/sync/struct.GILOnceCell.html#example-using-giloncecell-to-avoid-the-overhead-of-importing-a-class-multiple-times)).
```
Sure, just made those changes in https://github.com/PyO3/pyo3/pull/4542/commits/46b18cc0a56a52987dc5c009c0a68f809fbacda8.

However, since I changed the implementation from `PyType` to `T: PyTypeCheck`, this _does_ work:

```rust
#[pyfunction]
fn sum<'py>(py: Python<'py>, l: Py<PyList>) -> PyResult<Bound<'py, PyAny>> {
    static SUM: GILOnceCell<Py<PyAny>> = GILOnceCell::new();
    SUM.import(py, "builtins", "sum")?
        .call1((l,))
}
```

I guess this raises some questions. Is there any benefit in storing functions such as `sum` in the `GILOnceCell`, too? If not, should we prevent it (maybe by changing it back to `PyType`)? And if yes, should we be explicit in the docs that this is not only for importing classes?

mejrs reviewed (approved):
Thanks!
